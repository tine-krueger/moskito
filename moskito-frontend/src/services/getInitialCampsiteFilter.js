import { getFeatureBase } from './featureService'

export default function getInitialCampsiteFilter() {
    const initialFilter = 
        {
            name: 'Postleitzahl/Ort',
            dbName: 'postalCode',
            postalCode: '',
            features: getFeatureBase()
        }
    return initialFilter
}
