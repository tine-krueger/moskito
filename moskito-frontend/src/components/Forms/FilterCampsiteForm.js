import { useHistory } from 'react-router-dom'
import { useState } from 'react'
import styled from 'styled-components/macro'
import PropTypes from 'prop-types'
import Feature from './FilterCampsiteCheckbox'
import Suggestions from './Suggestions'
import Button from '../Button/Button'
import InputField from './InputField'
import useForm from '../../hooks/useForm'
import getInitialCampsiteFilter from '../../services/getInitialCampsiteFilter'
import { getGeocode, autocomplete} from '../../services/getGeocode'


FilterCampsiteForm.propTypes = {
    getCampsites: PropTypes.func.isRequired
}

export default function FilterCampsiteForm({getCampsites}) {
    const initialFilter = getInitialCampsiteFilter()
    const { fields, setValues, handleChange } = useForm(initialFilter) 
    const history = useHistory()
    const [ suggestions, setSuggestions ] = useState([])
    const [ errors, setErrors ] = useState()


    return (
        <FilterCampsite onSubmit={handleSubmit}>
            {errors && <p>{errors}</p>}
            <InputField data-testid='place' type={'text'} name={'postalCode'} value={fields.postalCode} onChange={handleLocationChange} placeholder={fields.name}>Dein Ziel:</InputField>
            {suggestions.length !== 0 && <Suggestions data-testid='suggestions' suggestions={suggestions} onClick={handleClick}/> }
            <InputField type={'number'} name={'distance'} value={fields.distance} onChange={handleChange} placeholder={fields.distance}>Umkreis in km:</InputField>
            <h3>(Keine) Ausstattung:</h3>
            <Checkboxes>
                {initialFilter.features.map(feature => 
                        <Feature key={feature.id} feature={feature} filter={fields} setFilter={setValues} />
                    )}
            </Checkboxes>  
            <Button>Campingpl채tze finden</Button>
        </FilterCampsite>
    )

    function handleSubmit(event){
        event.preventDefault()
        getCampsites(fields, setErrors)
        history.push('/campsites')
    }

    function handleClick(suggestion) {
        const place = suggestion.address.postalCode + " " + suggestion.address.city
        getGeocode(place)
        .then(result => {
            const geocodes = result.Response.View[0].Result[0].Location.DisplayPosition
            setValues(
                {
                    ...fields,
                    latitude: geocodes.Latitude,
                    longitude: geocodes.Longitude,
                    postalCode: place
                }
            )
        })
        .catch(() => 
            setSuggestions([
                { label: 'Ortsuche zu ungenau. Gib zus채tzlich eine PLZ ein oder versuche es zu einem sp채teren Zeitpunkt nocheinmal.'}
            ]));
        setSuggestions([])
    }

    function handleLocationChange(event) {
        setValues({
            ...fields,
            [event.target.name]: event.target.value
        })
        fields.postalCode.length > 2 ?
        autocomplete(fields.postalCode)
        .then(results => setSuggestions(results.suggestions)) 
        .catch(() =>
          setSuggestions([
            { label: 'Service nicht verf체gbar' },
          ])
        ) : setSuggestions([])
    }
}

const FilterCampsite = styled.form `
    max-width: 500px;
    display: grid;
    margin:2em 2em 3em 2em;

    h4 {
        padding-left: .5em;
        margin: 0 0 1em;
    }
`
const Checkboxes = styled.div`
    margin-top: 1em;
    padding-left: .5em;
`