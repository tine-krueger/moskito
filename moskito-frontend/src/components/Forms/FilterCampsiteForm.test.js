import React from 'react'
import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import 'jest-styled-components'
import FilterCampsiteForm from './FilterCampsiteForm'
import getInitialCampsiteFilter from '../../services/getInitialCampsiteFilter'
//import { useFilterForm } from '../../hooks/useFilterForm'
//import * as hooks from '../../hooks/useFilterForm'


const mockGetCampsites = jest.fn()
const mockHistoryPush = jest.fn()
const mockInputs = getInitialCampsiteFilter()
let mockSuggestions = []
const mockHandleCheckboxChange = jest.fn()
const mockHandleSubmit = jest.fn()
const mockHandleChange = jest.fn()
const mockHandleLocationChange = jest.fn()
const mockHandleClick = jest.fn()


jest.mock('../../hooks/useFilterForm.js', () => ({
    useFilterForm: () => ({
      inputs: mockInputs,
      suggestions: mockSuggestions,
      handleCheckboxChange: mockHandleCheckboxChange,
      handleSubmit: mockHandleSubmit,
      handleChange: mockHandleChange,
      handleLocationChange: mockHandleLocationChange,
      handleClick: mockHandleClick
    })
}))

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
      push: mockHistoryPush,
    }),
  }))

const component = <FilterCampsiteForm 
getCampsites={mockGetCampsites}
/>

describe('Filter Campsite Form', () => {
    it('renders correctly', () => {
      const { container } = render(component)
      expect(container).toMatchSnapshot()
    })
    
    it('shows all required input fields with initial values', () => {
      const { getByTestId } = render(component)
      expect(getByTestId('postalCode').value).toBe('')
      expect(getByTestId('distance').value).toEqual('100')
    })

    it('displays a button for submitting the form', () => {
      const { getByRole } = render(component)
      expect(getByRole('button', {name: 'Campingplätze finden' })).toBeInTheDocument()
    })

    it('calls ChangeHandlers, when user types something in', () => {
      const { getByTestId } = render(component)
      userEvent.type(getByTestId('postalCode'), 'Hamburg')
      userEvent.type(getByTestId('distance'), '50')
      expect(mockHandleLocationChange).toHaveBeenCalledTimes(7)
      expect(mockHandleChange).toHaveBeenCalledTimes(2)
    })

    it('calls SubmitHandler when clicking the Button', () => {
      mockHandleSubmit.mockImplementation(event => {
        event.preventDefault()
      })
      const { getByTestId, getByText } = render(component)
      userEvent.type(getByTestId('postalCode'), 'Hamburg')
      userEvent.type(getByTestId('distance'), '50')
      userEvent.click(getByText('Campingplätze finden'))
      expect(mockHandleSubmit).toHaveBeenCalledTimes(1)
      
    })

    it('shows suggestions, when they are not empty', () => {
      mockSuggestions = [{label: 'Place 1'}, {label: 'Place 2'}]
      const { getByText } = render(component)
      expect(getByText('Place 1')).toBeInTheDocument()
      expect(getByText('Place 2')).toBeInTheDocument()
    })

    
})