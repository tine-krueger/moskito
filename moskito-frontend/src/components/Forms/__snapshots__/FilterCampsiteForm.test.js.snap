// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Campsite Form renders correctly 1`] = `
.c3 {
  position: relative;
  height: 2.5em;
  margin-bottom: .75em;
}

.c3 input[type="checkbox"] {
  opacity: 0;
  position: absolute;
  visibility: hidden;
}

.c3 p {
  margin: 0;
  padding: .1em 0 0 2.5em;
  font-size: 1.1em;
}

.c3 p::before {
  content: " ";
  display: inline-block;
  position: absolute;
  width: 25px;
  height: 25px;
  left: 0;
  top: 0;
  -webkit-transition: all .2s ease;
  -webkit-transition: all .2s ease;
  transition: all .2s ease;
  border-radius: 5px;
  background-color: #d8e6e4;
  box-shadow: inset 4px 4px 6px 0 rgba(0,0,0,0.2),inset -3px -4px 6px 0 rgba(255,255,255,0.3);
}

.c3 p::after {
  content: " ";
  width: 14px;
  height: 14px;
  top: 6px;
  position: absolute;
  opacity: 0;
  background-color: #c97f63;
  font-weight: 900;
  border-radius: 2px;
  display: inline-block;
  left: 6px;
}

.c3 input[type="checkbox"]:checked+label>p::after {
  opacity: 1;
}

.c3 input[type="checkbox"]:checked+label>p::before {
  box-shadow: inset 5px 5px 3px -3px rgba(0,0,0,0.2),inset -3px -4px 6px 0 rgba(255,255,255,.9);
}

.c4 {
  padding: 1em 2em;
  font-family: 'Arvo';
  font-size: 1.1rem;
  border-radius: 10px;
  width: auto;
  border: none;
  background: #c2d6d3;
  box-shadow: 2px 4px 8px -4px rgba(0,0,0,0.6),-2px -2px 2px 0px rgba(255,255,255,0.5);
  color: #476c85;
  text-transform: uppercase;
  margin: 1em auto 3em auto;
}

.c4:hover {
  background: linear-gradient(145deg,#A2B2B0,#E2FAF6);
  box-shadow: 2px 4px 8px -4px rgba(0,0,0,0.6),-2px -2px 2px 0px rgba(255,255,255,0.5);
  outline: none;
}

.c1 {
  grid-column: 1 / -1;
  width: 100%;
  text-align: center;
  margin: 0 auto 2em;
  margin-bottom: 1em;
}

.c1 h3 {
  text-align: left;
  padding-left: .5em;
  margin: 0 0 1em;
}

.c1 input {
  width: 100%;
  padding: 1em;
  height: 3em;
  font-size: 1em;
  border: none;
  border-radius: 10px;
  color: #6b717e;
  background-color: #d8e6e4;
  box-shadow: inset 4px 4px 6px 0 rgba(0,0,0,0.2),inset -3px -4px 6px 0 rgba(255,255,255,0.3);
}

.c1 input:focus {
  outline: none;
  box-shadow: inset 4px 4px 6px 0 rgba(0,0,0,0.2),inset -3px -4px 6px 0 rgba(255,255,255,0.3),0 0 0px 1.7px #d3a392;
}

.c0 {
  max-width: 500px;
  display: grid;
  margin: 2em 2em 3em 2em;
}

.c0 h4 {
  padding-left: .5em;
  margin: 0 0 1em;
}

.c2 {
  margin-top: 1em;
  padding-left: .5em;
}

<div>
  <form
    class="c0"
  >
    <label
      class="c1"
      data-testid="input-field"
    >
      <h3>
        Dein Ziel:
      </h3>
      <input
        name="postalCode"
        placeholder="PLZ/Ort"
        type="text"
        value=""
      />
    </label>
    <label
      class="c1"
      data-testid="input-field"
    >
      <h3>
        Umkreis in km:
      </h3>
      <input
        name="distance"
        placeholder="100"
        type="number"
        value="100"
      />
    </label>
    <h3>
      (Keine) Ausstattung:
    </h3>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <input
          data-testid="wlan"
          id="wlan"
          type="checkbox"
        />
        <label
          for="wlan"
        >
          <p>
            kein W-Lan
          </p>
        </label>
      </div>
      <div
        class="c3"
      >
        <input
          data-testid="music"
          id="music"
          type="checkbox"
        />
        <label
          for="music"
        >
          <p>
            keine Bar/Diskothek
          </p>
        </label>
      </div>
      <div
        class="c3"
      >
        <input
          data-testid="animation"
          id="animation"
          type="checkbox"
        />
        <label
          for="animation"
        >
          <p>
            keine Animation
          </p>
        </label>
      </div>
      <div
        class="c3"
      >
        <input
          data-testid="fire"
          id="fire"
          type="checkbox"
        />
        <label
          for="fire"
        >
          <p>
            Lagerfeuerstelle
          </p>
        </label>
      </div>
      <div
        class="c3"
      >
        <input
          data-testid="path"
          id="path"
          type="checkbox"
        />
        <label
          for="path"
        >
          <p>
            Kies- oder Sandwege
          </p>
        </label>
      </div>
      <div
        class="c3"
      >
        <input
          data-testid="bulli"
          id="bulli"
          type="checkbox"
        />
        <label
          for="bulli"
        >
          <p>
            Bulli-Wiese
          </p>
        </label>
      </div>
      <div
        class="c3"
      >
        <input
          data-testid="tents"
          id="tents"
          type="checkbox"
        />
        <label
          for="tents"
        >
          <p>
            Zeltbereiche
          </p>
        </label>
      </div>
      <div
        class="c3"
      >
        <input
          data-testid="subdevision"
          id="subdevision"
          type="checkbox"
        />
        <label
          for="subdevision"
        >
          <p>
            keine Parzellierung
          </p>
        </label>
      </div>
      <div
        class="c3"
      >
        <input
          data-testid="permanent"
          id="permanent"
          type="checkbox"
        />
        <label
          for="permanent"
        >
          <p>
            wenig/keine Dauercamper
          </p>
        </label>
      </div>
      <div
        class="c3"
      >
        <input
          data-testid="size"
          id="size"
          type="checkbox"
        />
        <label
          for="size"
        >
          <p>
            mind. 80qm Standfläche
          </p>
        </label>
      </div>
      <div
        class="c3"
      >
        <input
          data-testid="bio"
          id="bio"
          type="checkbox"
        />
        <label
          for="bio"
        >
          <p>
            Bio/Regionale Lebensmittel
          </p>
        </label>
      </div>
      <div
        class="c3"
      >
        <input
          data-testid="snack"
          id="snack"
          type="checkbox"
        />
        <label
          for="snack"
        >
          <p>
            Imbiss
          </p>
        </label>
      </div>
      <div
        class="c3"
      >
        <input
          data-testid="animals"
          id="animals"
          type="checkbox"
        />
        <label
          for="animals"
        >
          <p>
            Hunde erlaubt
          </p>
        </label>
      </div>
      <div
        class="c3"
      >
        <input
          data-testid="seaside"
          id="seaside"
          type="checkbox"
        />
        <label
          for="seaside"
        >
          <p>
            Zugang zum See/Meer
          </p>
        </label>
      </div>
      <div
        class="c3"
      >
        <input
          data-testid="bathing"
          id="bathing"
          type="checkbox"
        />
        <label
          for="bathing"
        >
          <p>
            Badestelle für Kinder
          </p>
        </label>
      </div>
      <div
        class="c3"
      >
        <input
          data-testid="forest"
          id="forest"
          type="checkbox"
        />
        <label
          for="forest"
        >
          <p>
            mit Wald
          </p>
        </label>
      </div>
    </div>
    <button
      class="c4"
    >
      Campingplätze finden
    </button>
  </form>
</div>
`;
